stream/ResponseCollector.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package grakn.client.stream;
+
+import grakn.client.common.GraknClientException;
+import grakn.common.collection.Either;
+import io.grpc.StatusRuntimeException;
+
+import javax.annotation.Nullable;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import static grakn.client.common.ErrorMessage.Client.TRANSACTION_CLOSED;
+import static grakn.client.common.ErrorMessage.Internal.UNEXPECTED_INTERRUPTION;
+
+public class ResponseCollector<R> {
+
+    private final ConcurrentMap<UUID, Queue<R>> collectors;
+
+    public ResponseCollector() {
+        collectors = new ConcurrentHashMap<>();
+    }
+
+    public synchronized Queue<R> queue(UUID requestId) {
+        Queue<R> collector = new Queue<>();
+        collectors.put(requestId, collector);
+        return collector;
+    }
+
+    public Queue<R> get(UUID requestId) {
+        return collectors.get(requestId);
+    }
+
+    public synchronized void close(@Nullable StatusRuntimeException error) {
+        collectors.values().forEach(collector -> collector.close(error));
+        collectors.clear();
+    }
+
+    public static class Queue<R> {
+
+        private final BlockingQueue<Either<Response<R>, Done>> responseQueue;
```

Comment:
I think it might be cleaner to make `Response` a base class, with `Ok` and `Done` as subclasses - `Either` has a fairly unclear API and I also don't think it exists in many languages.

Reply:
let's keep it as discussed in person